prune.go

```diff
@@ -129,122 +135,194 @@ func txnsPruneC(txnsName string) string {
 			Queue []string `bson:"txn-queue"`
 		}
-		iter := coll.Find(nil).Select(bson.M{"txn-queue": 1}).Iter()
+		query := coll.Find(nil).Select(bson.M{"txn-queue": 1})
+		query.Batch(batchBytes)
+		iter := query.Iter()
 		for iter.Next(&tDoc) {
 			for _, token := range tDoc.Queue {
-				delete(txnIds, txnTokenToId(token))
+				if err := remover.remove(txnTokenToId(token)); err != nil {
+					return fmt.Errorf("handling completed txns: %v", err)
+				}
+				if t.isAfter() {
+					logger.Debugf("%d referenced txns found so far", remover.removed)
+				}
 			}
 		}
 		if err := iter.Close(); err != nil {
 			return fmt.Errorf("failed to read docs: %v", err)
 		}
 	}
+	if err := remover.flush(); err != nil {
+		return fmt.Errorf("handling completed txns: %v", err)
+	}
+	logger.Debugf("%d txns are still referenced and will be kept", remover.removed)
 
-	// Remove the unreferenced transactions.
-	logger.Debugf("%d transactions to remove", len(txnIds))
-	if err := bulkRemoveTxns(txns, txnIds); err != nil {
-		return fmt.Errorf("txn removal failed: %v", err)
+	// Remove the no-longer-referenced transactions from the txns collection.
+	t = newSimpleTimer(logInterval)
+	remover = newBulkRemover(txns)
+	defer remover.close()
+	query := workingSet.Find(nil).Batch(batchBytes)
+	iter := query.Iter()
+	var doc bson.D
+	for iter.Next(&doc) {
+		if err := remover.remove(doc[0].Value); err != nil {
+			return fmt.Errorf("removing txns: %v", err)
+		}
+		if t.isAfter() {
+			logger.Debugf("%d completed txns pruned so far", remover.removed)
+		}
+	}
+	if err := remover.flush(); err != nil {
+		return fmt.Errorf("removing txns: %v", err)
 	}
-	logger.Debugf("completed transactions pruned")
+
+	logger.Debugf("pruning completed: removed %d txns", remover.removed)
 	return nil
 }
 
+func runPipe(pipe *mgo.Pipe, result interface{}) error {
+	pipe.Batch(batchBytes)
+	pipe.AllowDiskUse()
+	return pipe.All(result)
+}
+
 func txnTokenToId(token string) bson.ObjectId {
 	// mgo/txn transaction tokens are the 24 character txn id
 	// followed by "_<nonce>"
 	return bson.ObjectIdHex(token[:24])
 }
 
-// bulkRemoveTxns removes transaction documents in chunks. It should
-// be significantly more efficient than removing one document per
-// remove query while also not trigger query document size limits.
-func bulkRemoveTxns(txns *mgo.Collection, txnIds map[bson.ObjectId]struct{}) error {
-	removeCount := 0
-	removeTxns := func(ids []bson.ObjectId) error {
-		bulk := txns.Bulk()
-		bulk.Unordered()
-		for _, id := range ids {
-			bulk.Remove(bson.D{{"_id", id}})
-		}
-		switch _, err := bulk.Run(); err {
-		case nil, mgo.ErrNotFound:
-			// It's OK for txns to no longer exist. Another process
-			// may have concurrently pruned them.
-			removeCount += len(ids)
-			logger.Tracef("%d completed transactions removed", removeCount)
-			return nil
-		default:
-			return err
-		}
+func newBulkRemover(coll *mgo.Collection) *bulkRemover {
+	session := coll.Database.Session.Copy()
```

Comment:
It doesn't look like the Copy was needed so I've removed it. That simplifies a few other aspects as well.